UI and create order changes required for app switch in Demo app

kgangineni opened the pull request:
### Summary of changes

1. changes required to add return_url, cancel_url in create order request body
2. send app_url when app switch is enabled in create order request body and setup token request body
3. UI changes to opt-in or opt-out of app switch before creating order

kgangineni committed:
⏺ Add OAuth2 client token fetching with suspend PayPalWebCheckoutClient

  - Implement FetchClientToken class for OAuth2 client credentials flow
  - Add headers parameter to APIRequest for Basic auth support
  - Convert PayPalWebCheckoutClient.start() to suspend function
  - Integrate token fetching before launching web checkout
  - Add comprehensive error handling for token fetch failures

kgangineni committed:
⏺ Implement app switch eligibility flow for PayPal checkout

  - GraphQL integration for PatchCcoWithAppSwitchEligibility
  - Integrate app switch eligibility, get redirect URL for launching checkout
  - Launch with redirect URL when available, fallback to web checkout
  - Create PayPalWebCheckoutCallback fun interface following CardApproveOrderCallback pattern
  - Add overloaded start() function accepting PayPalWebCheckoutCallback

kgangineni committed:
Refactor code to re-use constants in headers

kgangineni committed:
Merge remote-tracking branch 'origin/feature/app_switch' into BooleanToOptInAppSwitch

kgangineni committed:
- Improve null check in parsing response
- Move parsing response logic to PatchCCOWithAppSwitchEligibilityResponse class
- Move reading query and creating json object for request to PatchCcoWithAppSwitchEligibilityRequest class
- Improve test cases
- add paypalNativeAppInstalled param to graphQL request

kgangineni committed:
in-progress demo app changes

kgangineni committed:
⏺ Implement app switch logic for vault operations and fix test issues

  - Add app switch eligibility check for vault functionality
  - Create PayPalWebVaultCallback interface for vault operation callbacks
  - Extend PayPalWebLauncher with launchWithUrl() method supporting TokenType mapping

kgangineni committed:
Fix unit test for createLowScopedAccessToken()

kgangineni committed:
Merge branch 'BooleanToOptInAppSwitch' into demoAppChangesForAppSwitch

kgangineni committed:
Create Order modification, send minimal params required for app switch

sshropshire commented on Demo/src/main/java/com/paypal/android/ui/paypalwebvault/PayPalWebVaultView.kt:
```diff
@@ -28,7 +32,7 @@ import com.paypal.android.utils.UIConstants
 
 @Composable
-fun PayPalWebVaultView(viewModel: PayPalWebVaultViewModel = hiltViewModel()) {
+fun PayPalWebVaultView(viewModel: PayPalVaultViewModel = hiltViewModel()) {
```
```suggestion
fun PayPalVaultView(viewModel: PayPalVaultViewModel = hiltViewModel()) {
```
I like shortening the prefix. We'll have to look into doing this in the SDK as well. I think it makes sense to simplify the naming scheme as much as possible.

sshropshire commented on Demo/src/main/java/com/paypal/android/usecase/CreateOrderUseCase.kt:
```diff
@@ -42,6 +42,28 @@ class CreateOrderUseCase @Inject constructor(
+                val paypalJSON = JSONObject()
+                    .put("experience_context", experienceContextJSON)
+
+                val paymentSourceJSON = JSONObject()
+                    .put("paypal", paypalJSON)
+
+                orderRequest.put("payment_source", paymentSourceJSON)
```
It would be really cool to go full GSON. I think this portion of the JSON is one area where GSON support will be difficult. PPaaS schemas aren't the most GSON friendly.

sshropshire commented on Demo/src/main/java/com/paypal/android/usecase/CreateOrderUseCase.kt:
```diff
@@ -42,6 +42,28 @@ class CreateOrderUseCase @Inject constructor(
+                val experienceContextJSON = JSONObject()
+                    .put("return_url", SUCCESS_URL)
+                    .put("cancel_url", CANCEL_URL)
+                    .put("native-app", nativeAppJSON)
```
```suggestion
                    .put("native_app", nativeAppJSON)
```
qq: Is this not an underscore as well?

kgangineni replied:
it is underscore, corrected,

sshropshire commented on Demo/src/main/java/com/paypal/android/ui/paypalweb/PayPalCheckoutView.kt:
```diff
@@ -58,23 +61,40 @@ fun PayPalWebView(
 @Composable
-private fun Step1_CreateOrder(uiState: PayPalWebUiState, viewModel: PayPalWebViewModel) {
+fun Step1_WebOrNativeCheckoutHeader(viewModel: PayPalCheckoutViewModel) {
+    val uiState: PayPalUiState by viewModel.uiState.collectAsStateWithLifecycle()
```
Any reason we create another StateFlow here? I wonder if this could cause a double render. A lot of the demo app is inspired by this [NowInAndroid app](https://github.com/android/nowinandroid/tree/main). You can see here the `collectAsStateWithLifecycle()` methods are only called [at the top level.](https://github.com/android/nowinandroid/blob/main/feature/foryou/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/foryou/ForYouScreen.kt#L113)

kgangineni replied:
ya, it's not needed, corrected

kgangineni committed:
Merge remote-tracking branch 'origin/feature/app_switch' into demoAppChangesForAppSwitch

kgangineni committed:
minor formatting

kgangineni committed:
cleanup and minor refactoring
remove unnecessary code in parsing
simplify steps in order processing in demo app
move urls to common constants file

kgangineni committed:
fix static analysis

sshropshire reviewed (approved):
🔥
